### PR TITLE
fix save and import context issue

### PR DESCRIPTION
**Pull request by erich-wang**
## Description

fix the issue https://github.com/Azure/azure-powershell-pr/issues/497

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Accounts/Authentication.ResourceManager/AzureRmProfile.cs [hunk 1]
```diff
@@ -44,6 +44,8 @@ public class AzureRmProfile : IAzureContextContainer, IProfileOperations
 
         public IDictionary<string, IAzureContext> Contexts { get; set; } = new ConcurrentDictionary<string, IAzureContext>(StringComparer.CurrentCultureIgnoreCase);
 
+        private bool ShouldRefreshContextsFromCache { get; set; } = true;
```
Comment (isra-fel) [inline src/Accounts/Authentication.ResourceManager/AzureRmProfile.cs]:
```suggestion
        private bool _shouldRefreshContextsFromCache { get; set; } = true;
```

File: src/Accounts/Authentication/Authentication/Clients/SharedTokenCacheClientFactory.cs [hunk 2]
```diff
@@ -58,13 +72,15 @@ public override void RegisterCache(IClientApplicationBase client)
                         // replace the handler with the real one
                         var cacheHelper = GetCacheHelper(client.AppConfig.ClientId);
                         cacheHelper.RegisterCache(client.UserTokenCache);
+                        client.UserTokenCache.SetBeforeWrite(BeforeWriteNotification);
```
[the remark this answers is not in the document]
Comment (erich-wang) [reply src/Accounts/Authentication/Authentication/Clients/SharedTokenCacheClientFactory.cs]: After second thought, I think the safer way is to read token from file when token is needed. There're two benefits:
1. If possible, don't save token as plain text in memory to reduce security risk
2. If user starts one session and call Save-AzContext directly, the token is not set yet by using SetBeforeWrite, and may there're some other corner cases.

**Pull request merged**